pkg/shuffle/shuffle.go

```diff
@@ -0,0 +1,43 @@
+package shuffle
+
+import (
+	"github.com/numaproj/numaflow/pkg/isb"
+	"hash"
+	"hash/fnv"
+)
+
+// Shuffle shuffles messages among ISB
+type Shuffle struct {
+	bufferIdentifiers []string
+	hash              hash.Hash64
+}
+
+// NewShuffle accepts list of buffer identifiers(unique identifier of isb)
+// and returns new shuffle instance
+func NewShuffle(bufferIdentifiers []string) *Shuffle {
+	return &Shuffle{
+		bufferIdentifiers: bufferIdentifiers,
+		hash:              fnv.New64(),
+	}
+}
+
+// ShuffleMessages accepts list of isb messages and returns the mapping of isb to messages
+func (s *Shuffle) ShuffleMessages(messages []*isb.Message) map[string][]*isb.Message {
+	buffersCount := uint64(len(s.bufferIdentifiers))
+
+	// hash of the message key returns a unique hashValue
+	// mod of hashValue will decide which isb it will belong
+	hashMap := make(map[string][]*isb.Message)
+	for _, message := range messages {
+		hashValue := s.generateHash(message.Key)
+		hashValue = hashValue % buffersCount
```

Comment:
should we consider generating hash ranges (some pre-fixed multiple of no of isbs) and distributing the hash range? wondering if that will ensure more even distribution.

Reply:
probability would still remain the same, that is 1 / (number of isbs), I don't think it will make any difference